lib/pac55xx/ccs.c

```diff
@@ -0,0 +1,271 @@
+}
+void ccs_pll_set_indiv(uint32_t div) {
+	if (div <= 15 && div >= 1) {
+		CCSPLLCTL = (CCSPLLCTL & ~CCS_PLLCTL_PLLINDIV(CCS_PLLCTL_PLLINDIV_MASK)) | CCS_PLLCTL_PLLINDIV(div);
+	}
+}
+void ccs_pll_set_fbdiv(uint32_t div) {
+	if (div <= 16383 && div >= 4) {
+		CCSPLLCTL = (CCSPLLCTL & ~CCS_PLLCTL_PLLFBDIV(CCS_PLLCTL_PLLFBDIV_MASK)) | CCS_PLLCTL_PLLFBDIV(div);
+	}
+}
+void css_pll_config_enable(uint32_t indiv, uint32_t fbdiv, uint32_t outdiv) {
+	ccs_pll_disable();
+	ccs_pll_set_fbdiv(fbdiv);
+	ccs_pll_set_outdiv(outdiv);
+	ccs_pll_set_indiv(indiv);
+	ccs_pll_enable();
+	while (!ccs_pll_locked()) ; /* Wait for PLL lock ~500us */
+}
+uint32_t ccs_get_peripheral_clk_freq(uint32_t periph, uint32_t select) {
+	switch (periph) {
+		case ADC_BASE:
+			return ccs_sclk_frequency;
+		case I2C_BASE:       /* fall through */
+		case USARTA_BASE:    /* fall through */
+		case USARTB_BASE:    /* fall through */
+		case USARTC_BASE:    /* fall through */
+		case USARTD_BASE:    /* fall through */
+		case CAN_BASE:       /* fall through */
+		case GPTIMERA_BASE:  /* fall through */
+		case GPTIMERB_BASE:
+			return ccs_pclk_frequency;
+		case TIMERA_BASE:  /* fall through */
+		case TIMERB_BASE:  /* fall through */
+		case TIMERC_BASE:  /* fall through */
+		case TIMERD_BASE:
+			return (select == 0) ? ccs_pclk_frequency : ccs_aclk_frequency;
+		case MEMCTL_BASE:
+			return (select == 0) ? CCS_ROSC_FREQ : ccs_hclk_frequency;
+		case WWDT_BASE:
+			return (select == 0) ? ccs_frclk_frequency : CCS_ROSC_FREQ;
+		case RTC_BASE:
+			return ccs_frclk_frequency;
+		case CRC_BASE:  /* fall through */
+		case SYS_TICK_BASE:
+			return ccs_hclk_frequency;
+		default:
+			cm3_assert_not_reached();
+			break;
+	}
+	return 0; /* 0 if unsupported */
+}
+
+void ccs_reset_clocks(void) {
+	CCSCTL = CCS_CTL_LDOEN | CCS_CTL_ROSCEN |
+			CCS_CTL_PCLKEN | CCS_CTL_ACLKEN |
+			CCS_CTL_ADCCLKEN | CCS_CTL_STCLKSLPEN;
+	CCSPLLCTL = 0;
+}
+
+int ccs_configure_clocks(const struct ccs_clk_config *config) {
+	FLASHLOCK = FLASHLOCK_ALLOW_MEMCTL_WRITE;
+
+	// set safe defaults:
+	ccs_reset_clocks();
+	ccs_frclkmux_select(CCS_CTL_FRCLKMUXSEL_ROSC);
+	ccs_sclkmux_select_frclk();
+	MEMCTL &= ~MEMCTL_MCLKSEL; // select ROSCCLK
+
+	if (config->mem_cachedis) {
+		MEMCTL |= MEMCTL_CACHEDIS; // turn off Flash Cache
+	} else {
+		MEMCTL &= ~MEMCTL_CACHEDIS; // turn on Flash Cache
+	}
+
+	ccs_frclkmux_select(CCS_CTL_FRCLKMUXSEL_CLKREF); // switch frclk to 4MHz CLKREF
+
+	switch (config->frclk_source) {
+		case CCS_CTL_FRCLKMUXSEL_ROSC:
+			ccs_frclkmux_select(CCS_CTL_FRCLKMUXSEL_ROSC);
+			ccs_frclk_frequency = CCS_ROSC_FREQ;
+			break;
+		case CCS_CTL_FRCLKMUXSEL_CLKREF:
+			ccs_frclkmux_select(CCS_CTL_FRCLKMUXSEL_CLKREF);
+			ccs_frclk_frequency = CCS_CLKREF_FREQ;
+			break;
+		case CCS_CTL_FRCLKMUXSEL_EXTCLK:
+			if (config->extclk_frequency > CCS_EXTCLK_MAX_FREQ
+					|| config->extclk_frequency == 0) {
+				FLASHLOCK = 0;
```

Comment:
Suggest `#define FLASHLOCK_LOCKED` or similar instead of magic number 0...here and below.

Reply:
instead of 0, using FLASHLOCK_CLEAR